### ISSUE TITLE
Native macOS SmartCard pairing and PIN dialogues won't show up

### ISSUE THREAD

**Issue by deg0nz** [OpenSC/OpenSC#2605]
The problem seems to be tied to my machine/card combination somehow. I'm also not sure if this is a macOS problem or a problem with OpenSC.

As the title says, the native macOS SmartCard pairing and PIN dialogues don't show up on my system. We have configured other Macs with a similar setup and those machines work. I also tried my card on another Mac we configured with OpenSC and the PIN prompt and pairing UI show up there (we use the smartcard to authenticate with Cisco AnyConnect VPN, this is where I expect the PIN prompt). Edit: And other cards from the company also *don't* trigger PIN prompts on my machine.

OpenSC seems to work fine.
Firefox and Adobe Reader are able to access the certs on the card without any issues (But they use their own PIN prompts, not the native macOS one). I can sign PDFs with Reader for example.
The card also works for authentication on the company's Windows machines and with AnyConnect on those machines.

It seems that macOS itself has problems accessing data on the card.

Also, the necessary root and intermediate certs are installed in the system keychain (via macOS Server MDM).
My IT department is currently starting to integrate Macs into their environment, so there is not too much support from them regarding this issue.

**I'm happy to provide more detailed logs, please tell me what you need. 
(I just thought providing logs for every command that follows may be a bit much...)**

## What I've checked/tested so far

 - `pcsctest` - successful
 - `pkcs11-tool --login --test` - no errors
 - `pkcs11-tool --list-objects` - shows all the objects
 - `opensc-tool --name`- Atos CardOS
 - `system_profiler SPSmartCardsDataType` - shows all the certs
 - `sc_auth identities` - lists identities from the card
 - `sc_auth pairing_ui -s status` - enabled
 - `security list-smartcards` - lists the smartcard when inserted
 - `security smartcards token -l` - is empty
 - Clean install of macOS - issue still exists
 - Reset SMC - issue still exists
 - Another card reader - issue still exists

## Here's some errors I was able to find
(But searching for them didn't provide any useful help...)

### `sudo sc_auth pair -h <HASH> -u <USER>`

Shows this error:

```
Cannot pair user (public key hash: <HASH>. The operation couldn’t be completed. (TKAuthenticationHintsProvider error -11 - SmartCard algorithm initialisation error.)
```

### `sc_auth verifypin`

Shows this error:
```
Verifying PIN of SmartCard in reader OMNIKEY AG 3121 USB
ERROR: Unable to select card application AID {length = 11, bytes = 0x<OBFUSCATED>}
```

### `sc_auth pairing_ui`

Hangs

### `sc_auth pairing_ui -f`

Causes `ctkbind` to throw this message in `Console.app`: 

```
No RSA decryption key or ECDH key was found for token: org.opensc-project.mac.opensctoken.OpenSCTokenApp.OpenSCToken:0208c11a000f242a
```

### Inserting the card

Creates the same error for `ctkbind` as `sc_auth pairing_ui -f`.

I'd appreciate any help on this topic.

### OpenSC

OpenSC-0.22.0-rc1-74-gc902e199, rev: c902e199, commit-time: 2021-08-10 11:09:03 +0200

### OS

I have this issue since always. I startet testing smartcard stuff with Big Sur, updated to Monterey 12.6 now.

### Reader 

OMNIKEY AG 3121 USB

Disclosure: I already posted this on StackExchange and Apple Support Communities 2 weeks ago, but did not get any useful response. So I'm asking you guys here as a last resort.

**Comment by frankmorgner** [comment]
The above error message
```
Verifying PIN of SmartCard in reader OMNIKEY AG 3121 USB
ERROR: Unable to select card application AID {length = 11, bytes = 0x<OBFUSCATED>}
```
doesn't come from OpenSC, I believe. Could it be that you've installed a different a different token driver on this machine, which grabs the card? I think some `sc_auth` command should list the token driver(**s**) which support your card and which are installed on the machine.

**Comment by frankmorgner** [comment]
An other problem may be related to https://github.com/frankmorgner/OpenSCToken/issues/36#issuecomment-819655754. It has a suggested fix, which you may want to try

**Comment by deg0nz** [comment]
Hi @frankmorgner,

thanks for the hints! 

Regarding the `ERROR: Unable to select card application AID {length = 11, bytes = 0x<OBFUSCATED>}` issue: We re-tested this on another machine where the prompts are showing up and it didn't work there either. I'm pretty sure this is another issue.

I tested the nightly build that was mentioned in the thread of https://github.com/frankmorgner/OpenSCToken/issues/36#issuecomment-819655754, but it didn't fix the problem unfortunately.

**Comment by deg0nz** [comment]
Just a quick update.

I analyzed the situation some more and I believe I found the reason why some cards work on our Macs and some don't: 
It could be related to the CardOS version deployed on the cards. Cards with CardOS 4.4 work fine while the ones with CardOS 5.3 don't work.

It seems that our company's "Smartcard Certupdate Tool" also updates the CardOS version which could lead to this behaviour. I read in the wiki, that some features in CardOS 5 are not supported: https://github.com/OpenSC/OpenSC/wiki/Siemens-CardOS-M4

This would also explain the behaviour in #2624
